### PR TITLE
Implement CanvasGroup and CanvasItem clipping

### PR DESCRIPTION
**Pull request by reduz**
* Allows merging several 2D objects into a single draw operation
* Use current node to clip children nodes
* Further fixes to Vulkan barriers
* Changed font texture generation to white, fixes dark eges when blurred
* Other small misc fixes to backbuffer code.

mask
cgroup1
clip

### CODE REVIEW

File: servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp [hunk 1]
```diff
@@ -1338,12 +1354,46 @@ void RasterizerCanvasRD::canvas_render_items(RID p_to_render_target, Item *p_ite
...
+			}
+
+			ci->canvas_group_owner = nullptr; //must be cleared
+		}
+
+		if (ci == canvas_group_owner) {
+			_render_items(p_to_render_target, item_count, canvas_transform_inverse, p_light_list, true);
+			item_count = 0;
+
+			//Rect2i group_rect = ci->global_rect_cache;
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp:1382]: dead code

File: servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp [hunk 2]
```diff
@@ -1671,12 +1721,22 @@ void RasterizerCanvasRD::ShaderData::set_code(const String &p_code) {
...
 		case BLEND_MODE_MIX: {
+			/*
 			attachment.enable_blend = true;
-			attachment.alpha_blend_op = RD::BLEND_OP_ADD;
+			attachment.alpha_blend_op = RD::BLEND_OP_MAXIMUM;
 			attachment.color_blend_op = RD::BLEND_OP_ADD;
 			attachment.src_color_blend_factor = RD::BLEND_FACTOR_SRC_ALPHA;
 			attachment.dst_color_blend_factor = RD::BLEND_FACTOR_ONE_MINUS_SRC_ALPHA;
 			attachment.src_alpha_blend_factor = RD::BLEND_FACTOR_ONE;
+			attachment.dst_alpha_blend_factor = RD::BLEND_FACTOR_ONE;*/
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp:1731]: dead code

File: scene/main/canvas_item.h [hunk 3]
```diff
@@ -315,6 +316,9 @@ class CanvasItem : public Node {
 
 	void update();
 
+	void set_clip_children(bool p_visible);
```
Comment (akien-mga) [inline scene/main/canvas_item.h:319]:
```suggestion
	void set_clip_children(bool p_enable);
```

File: servers/rendering/rasterizer.h [hunk 4]
```diff
@@ -1060,6 +1061,17 @@ class RasterizerCanvas {
...
+		struct CanvasGroup {
+			RS::CanvasGroupMode mode;
+			bool fit_empty;
+			float fit_margin;
+			bool blur_mipmaps;
+			float clear_margin;
+		};
+
+		CanvasGroup *canvas_group = nullptr;
+		RS::CanvasGroupMode canvas_group_mode;
```
Comment (akien-mga) [inline servers/rendering/rasterizer.h:1074]: Seems like this might be unused? Probably replaced by `canvas_group.mode`?

File: servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp (hunk 1 above)
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp:1382]: Dead code.

File: servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp [hunk 5]
```diff
@@ -1671,12 +1721,22 @@ void RasterizerCanvasRD::ShaderData::set_code(const String &p_code) {
 
 		} break;
 		case BLEND_MODE_MIX: {
+			/*
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp:1724]: Is this commented code worth keeping? Old code will still be in the git history.

File: servers/rendering/rasterizer.h [hunk 6]
```diff
@@ -815,7 +815,8 @@ class RasterizerCanvas {
 		CANVAS_RECT_FLIP_H = 4,
 		CANVAS_RECT_FLIP_V = 8,
 		CANVAS_RECT_TRANSPOSE = 16,
-		CANVAS_RECT_CLIP_UV = 32
+		CANVAS_RECT_CLIP_UV = 32,
+		CANVAS_RECT_IS_GROUP = 64
```
Comment (akien-mga) [inline servers/rendering/rasterizer.h:819]: Note: If you take the habit to add a trailing comma to enums, you avoid this situation where a needless change is done on the previous last value.

File: servers/rendering/rasterizer_rd/rasterizer_effects_rd.h [hunk 7]
```diff
@@ -83,7 +85,8 @@ class RasterizerEffectsRD {
 		COPY_FLAG_FLIP_Y = (1 << 5),
 		COPY_FLAG_FORCE_LUMINANCE = (1 << 6),
 		COPY_FLAG_ALL_SOURCE = (1 << 7),
-		COPY_FLAG_HIGH_QUALITY_GLOW = (1 << 8)
+		COPY_FLAG_HIGH_QUALITY_GLOW = (1 << 8),
+		COPY_FLAG_ALPHA_TO_ONE = (1 << 9)
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_effects_rd.h:89]: Same here, if you add a trailing comma you won't get `-1, +2` next time but only `+1` (also nice for `git blame`).

File: servers/rendering/rasterizer_rd/rasterizer_storage_rd.cpp [hunk 8]
```diff
@@ -6283,21 +6300,30 @@ void RasterizerStorageRD::render_target_do_clear_request(RID p_render_target) {
...
+		region = Rect2i(Size2i(), rt->size).clip(p_region);
+		if (region.size == Size2i()) {
+			return; //nothing to do
+		}
 	}
 
 	//single texture copy for backbuffer
-	RD::get_singleton()->texture_copy(rt->color, rt->backbuffer_mipmap0, Vector3(region.position.x, region.position.y, 0), Vector3(region.position.x, region.position.y, 0), Vector3(region.size.x, region.size.y, 1), 0, 0, 0, 0, true);
-	//effects.copy(rt->color, rt->backbuffer_fb, blur_region);
+	//RD::get_singleton()->texture_copy(rt->color, rt->backbuffer_mipmap0, Vector3(region.position.x, region.position.y, 0), Vector3(region.position.x, region.position.y, 0), Vector3(region.size.x, region.size.y, 1), 0, 0, 0, 0, true);
```
Comment (akien-mga) [inline servers/rendering/rasterizer_rd/rasterizer_storage_rd.cpp:6321]: Is this relevant to keep?

File: servers/rendering_server.h [hunk 9]
```diff
@@ -1184,6 +1184,14 @@ class RenderingServer : public Object {
 
 	virtual void canvas_item_set_use_parent_material(RID p_item, bool p_enable) = 0;
 
+	enum CanvasGroupMode {
+		CANVAS_GROUP_MODE_DISABLED,
+		CANVAS_GROUP_MODE_OPAQUE,
+		CANVAS_GROUP_MODE_TRANSPARENT
```
Comment (akien-mga) [inline servers/rendering_server.h:1190]: Add trailing comma maybe?

**Review by akien-mga — APPROVED** [review]
I gave it a quick read and pinpointed some potential code quality fixes.
Giving it OK already anyway as it seems mostly good to merge, so it can be merged while I sleep ;)

File: servers/rendering/rasterizer_rd/rasterizer_storage_rd.cpp (hunk 8 above)
Comment (reduz) [reply servers/rendering/rasterizer_rd/rasterizer_storage_rd.cpp:6321]: yeah i guess, i replaced it but not fully trust it yet

File: scene/resources/dynamic_font.cpp [hunk 10]
```diff
@@ -458,8 +458,19 @@ DynamicFontAtSize::TexturePosition DynamicFontAtSize::_find_texture_pos_for_glyp
 			//zero texture
 			uint8_t *w = tex.imgdata.ptrw();
 			ERR_FAIL_COND_V(texsize * texsize * p_color_size > tex.imgdata.size(), ret);
-			for (int i = 0; i < texsize * texsize * p_color_size; i++) {
-				w[i] = 0;
+
```
Comment (Calinou) [inline scene/resources/dynamic_font.cpp:461]: It would be good to add a comment to explain why this is done, as it's not 100% obvious:

```suggestion
			// Initialize the texture to all-white pixels to prevent artifacts when the
			// font is displayed at a non-default scale with filtering.
			
```

(I'm adding this to my `3.2` backport.)

**Comment by akien-mga** [comment]
A "Update $filename" commit and it's not even Halloween yet 🙀

**Comment by Zireael07** [comment]
Vulkan only or is that for GLES too?

**Comment by Calinou** [comment]
@Zireael07 Given how this relies on low-level rendering changes, I doubt this can be reimplemented in GLES3 or GLES2 efficiently.
